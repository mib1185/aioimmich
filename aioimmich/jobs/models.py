"""aioimmich jobs models."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from mashumaro import field_options
from mashumaro.mixins.json import DataClassJSONMixin


class JobId(StrEnum):
    """Job IDs."""

    BACKGROUND_TASK = "backgroundTask"
    BACKUP_DATABASE = "backupDatabase"
    DUPLICATE_DETECTION = "duplicateDetection"
    FACE_DETECTION = "faceDetection"
    FACIAL_RECOGNITION = "facialRecognition"
    LIBRARY = "library"
    METADATA_EXTRACTION = "metadataExtraction"
    MIGRATION = "migration"
    NOTIFICATIONS = "notifications"
    SEARCH = "search"
    SIDECAR = "sidecar"
    SMART_SEARCH = "smartSearch"
    STORAGE_TEMPLATE_MIGRATION = "storageTemplateMigration"
    THUMBNAIL_GENERATION = "thumbnailGeneration"
    VIDEO_CONVERSION = "videoConversion"


class JobCommand(StrEnum):
    """Job commands."""

    CLEAR_FAILED = "clear-failed"
    EMPTY = "empty"
    PAUSE = "pause"
    RESUME = "resume"
    START = "start"


@dataclass
class JobCounts(DataClassJSONMixin):
    """Representation of job counts for a specific job type."""

    active: int
    completed: int
    delayed: int
    failed: int
    paused: int
    waiting: int


@dataclass
class QueueStatus(DataClassJSONMixin):
    """Representation of queue status for a specific job type."""

    is_active: bool = field(metadata=field_options(alias="isActive"))
    is_paused: bool = field(metadata=field_options(alias="isPaused"))


@dataclass
class ImmichJobStatus(DataClassJSONMixin):
    """Representation of status for a specific job type."""

    job_counts: JobCounts = field(metadata=field_options(alias="jobCounts"))
    queue_status: QueueStatus = field(metadata=field_options(alias="queueStatus"))


@dataclass
class ImmichAllJobStatus(DataClassJSONMixin):
    """Representation of all jobs status."""

    background_task: ImmichJobStatus = field(
        metadata=field_options(alias="backgroundTask")
    )
    backup_database: ImmichJobStatus = field(
        metadata=field_options(alias="backupDatabase")
    )
    duplicate_detection: ImmichJobStatus = field(
        metadata=field_options(alias="duplicateDetection")
    )
    face_detection: ImmichJobStatus = field(
        metadata=field_options(alias="faceDetection")
    )
    facial_recognition: ImmichJobStatus = field(
        metadata=field_options(alias="facialRecognition")
    )
    library: ImmichJobStatus
    metadata_extraction: ImmichJobStatus = field(
        metadata=field_options(alias="metadataExtraction")
    )
    migration: ImmichJobStatus
    notifications: ImmichJobStatus
    search: ImmichJobStatus
    sidecar: ImmichJobStatus
    smart_search: ImmichJobStatus = field(metadata=field_options(alias="smartSearch"))
    storage_template_migration: ImmichJobStatus = field(
        metadata=field_options(alias="storageTemplateMigration")
    )
    thumbnail_generation: ImmichJobStatus = field(
        metadata=field_options(alias="thumbnailGeneration")
    )
    video_conversion: ImmichJobStatus = field(
        metadata=field_options(alias="videoConversion")
    )
