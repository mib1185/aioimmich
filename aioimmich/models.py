"""aioimmich models."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class ImmichAsset:
    """Representation of an immich asset."""

    asset_id: str
    file_name: str
    mime_type: str


@dataclass
class ImmichAlbum:
    """Representation of an immich album."""

    album_id: str
    name: str
    description: str
    thumbnail_asset_id: str
    asset_count: int
    assets: list[ImmichAsset]
