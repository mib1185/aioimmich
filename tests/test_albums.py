"""Tests for aioimmich."""

from __future__ import annotations

from syrupy.assertion import SnapshotAssertion

from aioimmich.albums.models import ImmichAlbum


async def test_get_all_albums(mock_immich_with_data, snapshot: SnapshotAssertion):
    """Test async_get_all_albums."""
    api = await mock_immich_with_data()
    albums = await api.albums.async_get_all_albums()
    assert len(albums) == 2
    assert albums == snapshot


async def test_get_album_info(mock_immich_with_data, snapshot: SnapshotAssertion):
    """Test async_get_album_info."""
    api = await mock_immich_with_data()

    # get album info without assets
    album = await api.albums.async_get_album_info(
        "721e1a4b-aa12-441e-8d3b-5ac7ab283bb6", True
    )
    assert isinstance(album, ImmichAlbum)
    assert album == snapshot

    # get album info with assets
    album = await api.albums.async_get_album_info(
        "721e1a4b-aa12-441e-8d3b-5ac7ab283bb6", False
    )
    assert isinstance(album, ImmichAlbum)
    assert album == snapshot
