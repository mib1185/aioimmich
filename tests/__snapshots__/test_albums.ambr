# serializer version: 1
# name: test_get_album_info
  ImmichAlbum(album_id='721e1a4b-aa12-441e-8d3b-5ac7ab283bb6', album_name='Album 1', album_thumbnail_asset_id='0d03a7ad-ddc7-45a6-adee-68d322a6d2f5', album_users=[], asset_count=4, assets=[], created_at=datetime.datetime(2025, 5, 11, 10, 13, 22, 799000, tzinfo=datetime.timezone.utc), has_shared_link=False, is_activity_enabled=True, owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', owner=ImmichUser(avatar_color=<AvatarColor.PINK: 'pink'>, email='admin@myhost.local', name='admin', profile_changed_at=datetime.datetime(2025, 5, 11, 10, 7, 46, 866880, tzinfo=datetime.timezone.utc), profile_image_path='', user_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e'), shared=False, updated_at=datetime.datetime(2025, 5, 17, 11, 26, 3, 696000, tzinfo=datetime.timezone.utc), end_date=datetime.datetime(2025, 5, 13, 0, 0, tzinfo=datetime.timezone.utc), last_modified_asset_timestamp=datetime.datetime(2025, 5, 14, 19, 15, 27, 455000, tzinfo=datetime.timezone.utc), order='desc', start_date=datetime.datetime(2022, 11, 15, 0, 0, tzinfo=datetime.timezone.utc))
# ---
# name: test_get_album_info.1
  ImmichAlbum(album_id='721e1a4b-aa12-441e-8d3b-5ac7ab283bb6', album_name='Album 1', album_thumbnail_asset_id='0d03a7ad-ddc7-45a6-adee-68d322a6d2f5', album_users=[], asset_count=4, assets=[ImmichAsset(asset_id='2e94c203-50aa-4ad2-8e29-56dd74e0eff4', asset_type=<AssetType.IMAGE: 'IMAGE'>, checksum='HJm7TVOP80S+eiYZnAhWyRaB/Yc=', device_asset_id='web-20230131_182037.jpg-1675185639000', device_id='WEB', duration='0:00:00.00000', file_created_at=datetime.datetime(2023, 1, 31, 17, 20, 37, 85000, tzinfo=datetime.timezone.utc), file_modified_at=datetime.datetime(2023, 1, 31, 17, 20, 39, tzinfo=datetime.timezone.utc), has_metadata=True, is_archived=False, is_favorite=False, is_offline=False, is_trashed=False, local_datetime=datetime.datetime(2023, 1, 31, 18, 20, 37, 85000, tzinfo=datetime.timezone.utc), original_file_name='20230131_182037.jpg', original_path='upload/upload/e7ef5713-9dab-4bd4-b899-715b0ca4379e/b4/b8/b4b8ef00-8a6d-4056-91ff-7f86dc66e427.jpg', owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', people=[], resized=True, thumbhash='1igGFALX8mVGdHc5aChJf5nxNg==', updated_at=datetime.datetime(2025, 5, 11, 10, 13, 49, 590401, tzinfo=datetime.timezone.utc), duplicate_id=None, exif_info=ExifInfo(city=None, country=None, date_time_original=None, description=None, exif_image_height=None, exif_image_width=None, exposure_time=None, file_size_in_byte=None, f_number=None, focal_length=None, iso=None, latitude=None, lens_model=None, longitude=None, make=None, model=None, modify_date=None, orientation=None, projection_type=None, rating=None, state=None, time_zone=None), library_id=None, live_photo_video_id=None, original_mime_type='image/jpeg', visibility=None), ImmichAsset(asset_id='de104b75-6c7f-46a2-83ce-12a5b1c3a3b4', asset_type=<AssetType.IMAGE: 'IMAGE'>, checksum='aNCzp51jAuT3rXMv1icKLGMTC6M=', device_asset_id='web-20221222_205943.jpg-1671739185000', device_id='WEB', duration='0:00:00.00000', file_created_at=datetime.datetime(2022, 12, 22, 19, 59, 43, 95000, tzinfo=datetime.timezone.utc), file_modified_at=datetime.datetime(2022, 12, 22, 19, 59, 45, tzinfo=datetime.timezone.utc), has_metadata=True, is_archived=False, is_favorite=False, is_offline=False, is_trashed=False, local_datetime=datetime.datetime(2022, 12, 22, 20, 59, 43, 95000, tzinfo=datetime.timezone.utc), original_file_name='20221222_205943.jpg', original_path='upload/upload/e7ef5713-9dab-4bd4-b899-715b0ca4379e/43/ab/43abc8a2-3f81-4afc-95ea-c784cf869e4c.jpg', owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', people=[], resized=True, thumbhash='kWkGDAR9CGRpWO11VpJ2cGtVCQ==', updated_at=datetime.datetime(2025, 5, 11, 10, 13, 49, 590877, tzinfo=datetime.timezone.utc), duplicate_id=None, exif_info=ExifInfo(city=None, country=None, date_time_original=None, description=None, exif_image_height=None, exif_image_width=None, exposure_time=None, file_size_in_byte=None, f_number=None, focal_length=None, iso=None, latitude=None, lens_model=None, longitude=None, make=None, model=None, modify_date=None, orientation=None, projection_type=None, rating=None, state=None, time_zone=None), library_id=None, live_photo_video_id=None, original_mime_type='image/jpeg', visibility=None), ImmichAsset(asset_id='7d9ecb29-b12a-498d-adf9-c6ac09c761d3', asset_type=<AssetType.IMAGE: 'IMAGE'>, checksum='8fDGDOnOCVkJK8KiNu1bSmytEQ4=', device_asset_id='web-20221222_185242.jpg-1671731574000', device_id='WEB', duration='0:00:00.00000', file_created_at=datetime.datetime(2022, 12, 22, 17, 52, 44, 97000, tzinfo=datetime.timezone.utc), file_modified_at=datetime.datetime(2022, 12, 22, 17, 52, 54, tzinfo=datetime.timezone.utc), has_metadata=True, is_archived=False, is_favorite=False, is_offline=False, is_trashed=False, local_datetime=datetime.datetime(2022, 12, 22, 18, 52, 44, 97000, tzinfo=datetime.timezone.utc), original_file_name='20221222_185242.jpg', original_path='upload/upload/e7ef5713-9dab-4bd4-b899-715b0ca4379e/4f/a7/4fa7280f-16e6-4eb8-938a-89770266a20d.jpg', owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', people=[], resized=True, thumbhash='FUoGNAKHV4f4ZoLXZ1d2cl9H9w==', updated_at=datetime.datetime(2025, 5, 11, 10, 13, 49, 590479, tzinfo=datetime.timezone.utc), duplicate_id=None, exif_info=ExifInfo(city=None, country=None, date_time_original=None, description=None, exif_image_height=None, exif_image_width=None, exposure_time=None, file_size_in_byte=None, f_number=None, focal_length=None, iso=None, latitude=None, lens_model=None, longitude=None, make=None, model=None, modify_date=None, orientation=None, projection_type=None, rating=None, state=None, time_zone=None), library_id=None, live_photo_video_id=None, original_mime_type='image/jpeg', visibility=None), ImmichAsset(asset_id='1ed37a88-c720-49b9-bc3d-0e78aab3099a', asset_type=<AssetType.IMAGE: 'IMAGE'>, checksum='RA2WL5H511G+A7f1hV8+jJdfUUQ=', device_asset_id='web-20221115_185642.jpg-1668535006000', device_id='WEB', duration='0:00:00.00000', file_created_at=datetime.datetime(2022, 11, 15, 17, 56, 42, 77000, tzinfo=datetime.timezone.utc), file_modified_at=datetime.datetime(2022, 11, 15, 17, 56, 46, tzinfo=datetime.timezone.utc), has_metadata=True, is_archived=False, is_favorite=False, is_offline=False, is_trashed=False, local_datetime=datetime.datetime(2022, 11, 15, 18, 56, 42, 77000, tzinfo=datetime.timezone.utc), original_file_name='20221115_185642.jpg', original_path='upload/upload/e7ef5713-9dab-4bd4-b899-715b0ca4379e/7c/86/7c8643f0-f91e-4926-8e6a-8e8c4403e4d0.jpg', owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', people=[], resized=True, thumbhash='UBgODIJvdKmYZmZ2Z3cPlfUjaQ==', updated_at=datetime.datetime(2025, 5, 11, 10, 13, 49, 766801, tzinfo=datetime.timezone.utc), duplicate_id=None, exif_info=ExifInfo(city=None, country=None, date_time_original=None, description=None, exif_image_height=None, exif_image_width=None, exposure_time=None, file_size_in_byte=None, f_number=None, focal_length=None, iso=None, latitude=None, lens_model=None, longitude=None, make=None, model=None, modify_date=None, orientation=None, projection_type=None, rating=None, state=None, time_zone=None), library_id=None, live_photo_video_id=None, original_mime_type='image/jpeg', visibility=None)], created_at=datetime.datetime(2025, 5, 11, 10, 13, 22, 799000, tzinfo=datetime.timezone.utc), has_shared_link=False, is_activity_enabled=True, owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', owner=ImmichUser(avatar_color=<AvatarColor.PINK: 'pink'>, email='admin@myhost.local', name='admin', profile_changed_at=datetime.datetime(2025, 5, 11, 10, 7, 46, 866880, tzinfo=datetime.timezone.utc), profile_image_path='', user_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e'), shared=False, updated_at=datetime.datetime(2025, 5, 17, 11, 26, 3, 696000, tzinfo=datetime.timezone.utc), end_date=datetime.datetime(2025, 5, 13, 0, 0, tzinfo=datetime.timezone.utc), last_modified_asset_timestamp=datetime.datetime(2025, 5, 14, 19, 15, 27, 455000, tzinfo=datetime.timezone.utc), order='desc', start_date=datetime.datetime(2022, 11, 15, 0, 0, tzinfo=datetime.timezone.utc))
# ---
# name: test_get_all_albums
  list([
    ImmichAlbum(album_id='721e1a4b-aa12-441e-8d3b-5ac7ab283bb6', album_name='Album 1', album_thumbnail_asset_id='0d03a7ad-ddc7-45a6-adee-68d322a6d2f5', album_users=[], asset_count=4, assets=[], created_at=datetime.datetime(2025, 5, 11, 10, 13, 22, 799000, tzinfo=datetime.timezone.utc), has_shared_link=False, is_activity_enabled=True, owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', owner=ImmichUser(avatar_color=<AvatarColor.PINK: 'pink'>, email='admin@myhost.local', name='admin', profile_changed_at=datetime.datetime(2025, 5, 11, 10, 7, 46, 866880, tzinfo=datetime.timezone.utc), profile_image_path='', user_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e'), shared=False, updated_at=datetime.datetime(2025, 5, 17, 11, 26, 3, 696000, tzinfo=datetime.timezone.utc), end_date=datetime.datetime(2025, 5, 13, 0, 0, tzinfo=datetime.timezone.utc), last_modified_asset_timestamp=datetime.datetime(2025, 5, 14, 19, 15, 27, 455000, tzinfo=datetime.timezone.utc), order='desc', start_date=datetime.datetime(2022, 11, 15, 0, 0, tzinfo=datetime.timezone.utc)),
    ImmichAlbum(album_id='c4721c25-8669-451f-a6be-3d9d4dc38f5f', album_name='Album 2', album_thumbnail_asset_id='6961329c-d8ee-4f09-9d4e-6997ea5b7bf4', album_users=[], asset_count=426, assets=[], created_at=datetime.datetime(2025, 5, 12, 20, 48, 30, 784000, tzinfo=datetime.timezone.utc), has_shared_link=False, is_activity_enabled=True, owner_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e', owner=ImmichUser(avatar_color=<AvatarColor.PINK: 'pink'>, email='admin@myhost.local', name='admin', profile_changed_at=datetime.datetime(2025, 5, 11, 10, 7, 46, 866880, tzinfo=datetime.timezone.utc), profile_image_path='', user_id='e7ef5713-9dab-4bd4-b899-715b0ca4379e'), shared=False, updated_at=datetime.datetime(2025, 5, 17, 11, 34, 53, 942000, tzinfo=datetime.timezone.utc), end_date=datetime.datetime(2025, 5, 8, 0, 0, tzinfo=datetime.timezone.utc), last_modified_asset_timestamp=datetime.datetime(2025, 5, 17, 11, 34, 54, 77000, tzinfo=datetime.timezone.utc), order='desc', start_date=datetime.datetime(2017, 4, 11, 0, 0, tzinfo=datetime.timezone.utc)),
  ])
# ---
